pkg/tsdb/postgres/tlsmanager_test.go

```diff
@@ -0,0 +1,291 @@
+// +build integration
+
+package postgres
+
+import (
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/grafana/grafana/pkg/components/securejsondata"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	_ "github.com/lib/pq"
+)
+
+var writeCertFileCallNum int
+
+// TestDataSourceCacheManager is to test the Cache manager
+func TestDataSourceCacheManager(t *testing.T) {
+	cfg := setting.NewCfg()
+	cfg.DataPath = t.TempDir()
+	mng := tlsManager{
+		logger:          log.New("tsdb.postgres"),
+		dsCacheInstance: datasourceCacheManager{locker: newLocker()},
+		dataPath:        cfg.DataPath,
+	}
+
+	settings := tlsSettings{}
+	jsonData := `{"sslmode" : "verify-full", "tlsConfigurationMethod" : "file-content"}`
+	jsonDataValue, err := simplejson.NewJson([]byte(jsonData))
+	require.NoError(t, err)
+
+	secureJsonData := `{"tlsClientCert" : "I am client certification", "tlsClientKey" : "I am client key", "tlsCACert" : "I am CA certification"}`
+	securityjsonData := map[string]string{}
+	err = json.Unmarshal([]byte(secureJsonData), &securityjsonData)
+	securityjsonValue := securejsondata.GetEncryptedJsonData(securityjsonData)
+	require.NoError(t, err)
+
+	mockValidateCertFilePaths()
+	t.Cleanup(resetValidateCertFilePaths)
+
+	t.Run("Check datasource cache creation", func(t *testing.T) {
+		var id, index int64
+		var wg sync.WaitGroup
+		wg.Add(10)
+		mutex := new(sync.Mutex)
+		index = 1
+		for id = 1; id <= 10; id++ {
+			go func() {
+				mutex.Lock()
```

Comment:
Why is a mutex used here? Doesn't look like it's needed.

Reply:
for the index we talked about, I would remove the index as you suggested, and use back id as datasource id, then i would not need the mutex anymore

Reply:
Yes, in my head that should work :+1: